File: client/android/core/src/test/java/io/github/pstanar/pstotp/core/crypto/IconLibraryCryptoTest.kt
package io.github.pstanar.pstotp.core.crypto

import io.github.pstanar.pstotp.core.model.IconLibraryBlob
import io.github.pstanar.pstotp.core.model.LibraryIcon
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotEquals
import org.junit.Assert.assertThrows
import org.junit.Test
import java.security.SecureRandom

class IconLibraryCryptoTest {

    private fun randomKey(): ByteArray = ByteArray(32).also { SecureRandom().nextBytes(it) }

    private val sample = IconLibraryBlob(
        version = 1,
        icons = listOf(
            LibraryIcon(
                id = "11111111-1111-1111-1111-111111111111",
                label = "Company logo",
                data = "data:image/png;base64,iVBORw0KGgo=",
                createdAt = "2026-04-22T00:00:00.000Z",
            ),
            LibraryIcon(
                id = "22222222-2222-2222-2222-222222222222",
                label = "Another",
                data = "data:image/png;base64,XYZ=",
                createdAt = "2026-04-22T01:00:00.000Z",
            ),
        ),
    )

    @Test
    fun `roundtrip preserves icons and version`() {
        val key = randomKey()
        val encrypted = IconLibraryCrypto.encrypt(key, sample)
        val decrypted = IconLibraryCrypto.decrypt(key, encrypted)

        assertEquals(sample.version, decrypted.version)
        assertEquals(sample.icons.size, decrypted.icons.size)
        assertEquals(sample.icons[0], decrypted.icons[0])
        assertEquals(sample.icons[1], decrypted.icons[1])
    }

    @Test
    fun `fresh nonce every call — same input yields different ciphertext`() {
        val key = randomKey()
        val a = IconLibraryCrypto.encrypt(key, sample)
        val b = IconLibraryCrypto.encrypt(key, sample)
        assertNotEquals("Two encryptions of the same blob should differ (fresh nonce)",
            a.toList(), b.toList())
    }

    @Test
    fun `decrypt with wrong key fails the MAC`() {
        val encrypted = IconLibraryCrypto.encrypt(randomKey(), sample)
        // Any javax.crypto failure is fine; Assert.assertThrows covers them all.
        assertThrows(Exception::class.java) {
            IconLibraryCrypto.decrypt(randomKey(), encrypted)
        }
    }

    @Test
    fun `empty library roundtrips cleanly`() {
        val key = randomKey()
        val empty = IconLibraryBlob(version = 1, icons = emptyList())
        val decrypted = IconLibraryCrypto.decrypt(key, IconLibraryCrypto.encrypt(key, empty))
        assertEquals(0, decrypted.icons.size)
    }
}

File: client/web/src/tests/IconLibraryCrypto.test.ts
import { describe, it, expect } from "vitest";
import {
  encryptIconLibrary,
  decryptIconLibrary,
} from "@/features/vault/utils/icon-library-crypto";
import type { IconLibraryBlob } from "@/types/icon-library";
import { generateVaultKey } from "@/lib/crypto";

describe("icon-library crypto", () => {
  const sampleBlob: IconLibraryBlob = {
    version: 1,
    icons: [
      {
        id: "11111111-1111-1111-1111-111111111111",
        label: "Company logo",
        data: "data:image/png;base64,iVBORw0KGgo=",
        createdAt: "2026-04-22T00:00:00.000Z",
      },
      {
        id: "22222222-2222-2222-2222-222222222222",
        label: "Another",
        data: "data:image/png;base64,XYZ=",
        createdAt: "2026-04-22T01:00:00.000Z",
      },
    ],
  };

  it("round-trips a blob through encrypt + decrypt", async () => {
    const key = generateVaultKey();
    const encrypted = await encryptIconLibrary(key, sampleBlob);
    expect(encrypted).toMatch(/^[A-Za-z0-9+/=]+$/); // base64

    const decrypted = await decryptIconLibrary(key, encrypted);
    expect(decrypted).toEqual(sampleBlob);
  });

  it("produces different ciphertext each call (fresh nonce)", async () => {
    const key = generateVaultKey();
    const a = await encryptIconLibrary(key, sampleBlob);
    const b = await encryptIconLibrary(key, sampleBlob);
    expect(a).not.toEqual(b);
  });

  it("decrypting with a different key throws", async () => {
    const key1 = generateVaultKey();
    const key2 = generateVaultKey();
    const encrypted = await encryptIconLibrary(key1, sampleBlob);
    await expect(decryptIconLibrary(key2, encrypted)).rejects.toThrow();
  });

  it("handles the empty-library case", async () => {
    const key = generateVaultKey();
    const empty: IconLibraryBlob = { version: 1, icons: [] };
    const encrypted = await encryptIconLibrary(key, empty);
    const decrypted = await decryptIconLibrary(key, encrypted);
    expect(decrypted).toEqual(empty);
  });
});
